onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -72,6 +72,111 @@ bool ReshapeFusion::Fuse_Subgraph2(Graph& graph, const NodeArg& root_input, cons
+      }
+      return false;
+    }
+    const Node& binary_node = edges[1]->GetNode();
+
+    // Check if each of two inputs of the binary node has exactly one element.
+    auto input_count = binary_node.InputArgCount().front();
+
+    for (int i = 0; i < input_count; ++i) {
+      // Check if the i-th argument is a constant initializer. If it is, check for its shape
+      // to make sure it contains exactly one element.
+      if (optimizer_utils::AppendTensorFromInitializer(graph, *(binary_node.InputDefs()[i]), shape_value, true)) {
+        const NodeArg* binary_node_arg = binary_node.InputDefs()[i];
+        const auto* input_shape = binary_node_arg->Shape();
+        if (!input_shape) {
+          // We need shape to be able to be certain of number of elements
+          // Can't proceed with fusion
+          return false;
+        }
+
+        // Check if number of elements in this input to perform binary operation is 1
+        if (utils::GetTensorShapeFromTensorShapeProto(*binary_node_arg->Shape()).Size() != 1) {
+          // Some dim values may be > 1 or some dim values may be missing
+          // Can't proceed with fusion
+          return false;
+        }
+        continue;
+      }
+      // If it's node input, look for shape -> slice -> squeeze path for a potential match.
+      if (!ReshapeFusion::Match_One_Element_Output_Subgraph(graph, root_input, binary_node, i, logger)) {
+        return false;
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Find the subgraph that matches [root] -> Shape -> Slice -> Squeeze. Check the inputs of slice 
+ * to make sure the graph produces output with exactly one element.
+ */
+bool ReshapeFusion::Match_One_Element_Output_Subgraph(Graph& graph, const NodeArg& root_input, const Node& cur_node,
+                                                      int index, const logging::Logger& logger) {
+  std::vector<graph_utils::EdgeEndToMatch> parent_path{
+      {0, index, "Squeeze", {1, 11}, kOnnxDomain},
+      {0, 0, "Slice", {1, 11}, kOnnxDomain},
+      {0, 0, "Shape", {1}, kOnnxDomain}};
+  std::vector<const Node::EdgeEnd*> edges;
+  if (graph_utils::FindPath(cur_node, true, parent_path, edges, logger)) {
+    const Node& slice = edges[1]->GetNode();
+    const Node& shape = edges[2]->GetNode();
+
+    const NodeArg& shape_input = *(shape.InputDefs()[0]);
+    if (shape_input.Name() != root_input.Name()) {
+      return false;
+    }
+
+    // Check if Slice op slices 1d array (result of shape) to one element.
+    std::vector<int64_t> slice_inputs;
+    if (optimizer_utils::AppendTensorFromInitializer(graph, *(slice.InputDefs()[1]), slice_inputs, true) &&
+        optimizer_utils::AppendTensorFromInitializer(graph, *(slice.InputDefs()[2]), slice_inputs, true)) {
+      const int64_t slice_start = slice_inputs[0];
+      const int64_t slice_end = slice_inputs[1];
+      if (!(slice_end == INT64_MAX && slice_start == -1) && abs(slice_end - slice_start) != 1) {
```

Comment:
Not sure everyone uses INT64_MAX. It might be safe to check that slice_end larger than a large enough threshold.

Reply:
Right. Although he slice documentation states "For slicing to the end of a dimension with unknown size, it is recommended to pass in INT_MAX when slicing forward and 'INT_MIN' when slicing backward.", the GPT2 models uses INT64_MAX. I'll change the condition to > INT_MAX then.